exchanges/binance/binance.go

```diff
@@ -161,27 +162,89 @@ func (b *Binance) GetHistoricalTrades(symbol string, limit int, fromID int64) ([
-	var resp []AggregatedTrade
-
-	if err := b.CheckLimit(limit); err != nil {
-		return resp, err
-	}
-
+// GetAggregatedTrades returns aggregated trade activity.
+// If more than one hour of data is requested with no limit, the trades are collected with multiple backend requests.
+// https://binance-docs.github.io/apidocs/spot/en/#compressed-aggregate-trades-list
+func (b *Binance) GetAggregatedTrades(arg *AggregatedTradeRequestParams) ([]AggregatedTrade, error) {
 	params := url.Values{}
-	params.Set("symbol", strings.ToUpper(symbol))
-	if limit > 0 {
-		params.Set("limit", strconv.Itoa(limit))
+	symbol := arg.Symbol
+	symbol.Delimiter = ""
+	params.Set("symbol", strings.ToUpper(symbol.String()))
+	if arg.Limit > 0 {
+		if err := b.CheckLimit(arg.Limit); err != nil {
+			return nil, err
+		}
+		params.Set("limit", strconv.Itoa(arg.Limit))
+	}
+	if arg.FromID != 0 {
+		params.Set("fromId", strconv.FormatInt(arg.FromID, 10))
+	}
+	if !arg.StartTime.IsZero() {
+		params.Set("startTime", strconv.FormatInt(convert.UnixMillis(arg.StartTime), 10))
+	}
+	if !arg.EndTime.IsZero() {
+		params.Set("endTime", strconv.FormatInt(convert.UnixMillis(arg.EndTime), 10))
+	}
+	// If startTime and endTime are sent, time between startTime and endTime must be less than 1 hour.
+	if arg.Limit == 0 && !arg.StartTime.IsZero() && !arg.EndTime.IsZero() && arg.EndTime.Sub(arg.StartTime) > time.Hour {
+		// Split the request into multiple
+		return batchAggregateTrades(b, arg, params)
 	}
 
+	var resp []AggregatedTrade
 	path := b.API.Endpoints.URL + aggregatedTrades + "?" + params.Encode()
 	return resp, b.SendHTTPRequest(path, limitDefault, &resp)
 }
 
+// batchAggregateTrades fetches trades in multiple requests
+// first phase, hourly requests until the first trade (or end time) is reached
+// second phase, limit requests from previous trade until end time is reached
+func batchAggregateTrades(b *Binance, arg *AggregatedTradeRequestParams, params url.Values) ([]AggregatedTrade, error) {
+	var resp []AggregatedTrade
+	// prepare first request with only first hour and max limit
+	params.Set("limit", "1000")
+
+	for start := arg.StartTime; len(resp) == 0; start = start.Add(time.Hour) {
+		if !start.Before(arg.EndTime) {
+			// All requests returned empty
+			return nil, nil
+		}
+		params.Set("startTime", strconv.FormatInt(convert.UnixMillis(start), 10))
+		params.Set("endTime", strconv.FormatInt(convert.UnixMillis(start.Add(time.Hour)), 10))
+		path := b.API.Endpoints.URL + aggregatedTrades + "?" + params.Encode()
+		err := b.SendHTTPRequest(path, limitDefault, &resp)
+		if err != nil {
+			log.Warn(log.ExchangeSys, err.Error())
+			return resp, err
+		}
+	}
+
+	// other requests follow from the last aggregate trade id and have no time window
+	params.Del("startTime")
+	params.Del("endTime")
+	for {
+		// Keep requesting new data after last retrieved trade
+		params.Add("fromId", strconv.FormatInt(resp[len(resp)-1].ATradeID, 10))
```

Comment:
params.Add adds a new fromId to a map causing it to set 2 fromIds. either delete the previous id or params.Set() which i think overwrites it (atleast i think so):

 https://api.binance.com/api/v3/aggTrades?fromId=412630625&fromId=412631624&limit=1000&symbol=BTCUSDT
Binance request type: GET
    c:\dev\gocryptotrader\exchanges\binance\binance_test.go:104: Binance GetAggregatedTrades() error Binance unsuccessful HTTP status code: 500 raw response: {"code":-1000,"msg":"An unknown error occured while processing the request."}

Reply:
Yes, sorry... fixed...